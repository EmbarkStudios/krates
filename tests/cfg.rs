mod util;

use krates::cfg_expr::targets;
use util::{build, cmp};

#[test]
fn ignores_non_linux() {
    let mut kb = krates::Builder::new();
    kb.include_targets(targets::ALL_BUILTINS.iter().filter_map(|ti| {
        if ti.os == Some(targets::Os::linux) {
            Some((ti.triple.clone(), Vec::new()))
        } else {
            None
        }
    }));

    let grafs = build("all-features.json", kb).unwrap();

    let targets: Vec<_> = targets::ALL_BUILTINS
        .iter()
        .filter(|ti| ti.os == Some(targets::Os::linux))
        .collect();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_tier1() {
    let mut kb = krates::Builder::new();

    let targets = vec![
        targets::get_builtin_target_by_triple("i686-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("i686-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("i686-unknown-linux-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-apple-darwin").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-gnu").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-pc-windows-msvc").unwrap(),
        targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu").unwrap(),
    ];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[test]
fn ignores_non_wasm() {
    let mut kb = krates::Builder::new();

    let targets = vec![targets::get_builtin_target_by_triple("wasm32-unknown-unknown").unwrap()];

    kb.include_targets(targets.iter().map(|ti| (ti.triple.clone(), vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    for ti in &targets {
                        if expr.eval(|pred| match pred {
                            krates::cfg_expr::Predicate::Target(tp) => tp.matches(*ti),
                            _ => false,
                        }) {
                            println!("{cfg} matched {}", ti.triple);
                            return false;
                        }
                    }

                    true
                } else {
                    !targets.iter().any(|ti| ti.triple.as_str() == cfg)
                }
            } else {
                false
            }
        },
    );
}

#[cfg(feature = "targets")]
#[test]
fn handles_non_builtin() {
    let mut kb = krates::Builder::new();

    use krates::target_lexicon::Triple;

    let xbox: Triple = "x86_64-xboxone-windows-msvc".parse().unwrap();

    kb.include_targets(std::iter::once(("x86_64-xboxone-windows-msvc", vec![])));

    let grafs = build("all-features.json", kb).unwrap();

    cmp(
        grafs,
        |_| false,
        |ef| {
            if let Some(cfg) = ef.dep.and_then(|d| d.cfg) {
                if cfg.starts_with("cfg(") {
                    let expr = krates::cfg_expr::Expression::parse(cfg).unwrap();

                    !expr.eval(|pred| match pred {
                        krates::cfg_expr::Predicate::Target(tp) => tp.matches(&xbox),
                        _ => false,
                    })
                } else {
                    "x86_64-xboxone-windows-msvc" != cfg
                }
            } else {
                false
            }
        },
    );
}
