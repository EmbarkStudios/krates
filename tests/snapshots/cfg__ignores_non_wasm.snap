---
source: tests/cfg.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///home/jake/code/krates/tests/ws/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws/b)" ]
    2 [ label = "crate bumpalo 3.11.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate c 0.1.0 (path+file:///home/jake/code/krates/tests/ws/c)" ]
    4 [ label = "crate cc 1.0.73 (git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6)" ]
    5 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate js-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate leftpad 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate proc-macro2 1.0.44 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate syn 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate wasm-bindgen 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate wasm-bindgen-backend 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate wasm-bindgen-macro 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate wasm-bindgen-macro-support 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate wasm-bindgen-shared 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate web-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "feature default" ]
    22 [ label = "feature default" ]
    23 [ label = "feature Crypto" ]
    24 [ label = "feature Window" ]
    25 [ label = "feature default" ]
    26 [ label = "feature proc-macro" ]
    27 [ label = "feature proc-macro" ]
    28 [ label = "feature spans" ]
    29 [ label = "feature default" ]
    30 [ label = "feature default" ]
    31 [ label = "feature default" ]
    32 [ label = "feature default" ]
    33 [ label = "feature default" ]
    34 [ label = "feature full" ]
    35 [ label = "feature spans" ]
    36 [ label = "feature visit" ]
    37 [ label = "feature spans" ]
    38 [ label = "feature leftpad" ]
    39 [ label = "feature leftier-strings" ]
    40 [ label = "feature lazy_static" ]
    41 [ label = "feature std" ]
    42 [ label = "feature alloc" ]
    43 [ label = "feature race" ]
    44 [ label = "feature derive" ]
    45 [ label = "feature parsing" ]
    46 [ label = "feature printing" ]
    47 [ label = "feature clone-impls" ]
    48 [ label = "feature proc-macro" ]
    49 [ label = "feature quote" ]
    50 [ label = "feature spans" ]
    51 [ label = "feature std" ]
    52 [ label = "feature EventTarget" ]
    0 -> 1 [ label = "" ]
    0 -> 21 [ label = "(dev)" ]
    1 -> 21 [ label = "" ]
    3 -> 4 [ label = "(build)" ]
    3 -> 22 [ label = "(dev)" ]
    3 -> 8 [ label = "" ]
    3 -> 23 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    3 -> 24 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    7 -> 25 [ label = "" ]
    9 -> 5 [ label = "" ]
    11 -> 14 [ label = "" ]
    12 -> 11 [ label = "" ]
    12 -> 26 [ label = "" ]
    13 -> 11 [ label = "" ]
    13 -> 26 [ label = "" ]
    13 -> 12 [ label = "" ]
    13 -> 27 [ label = "" ]
    13 -> 14 [ label = "" ]
    15 -> 5 [ label = "" ]
    15 -> 17 [ label = "" ]
    15 -> 28 [ label = "" ]
    16 -> 29 [ label = "" ]
    16 -> 9 [ label = "" ]
    16 -> 30 [ label = "" ]
    16 -> 31 [ label = "" ]
    16 -> 32 [ label = "" ]
    16 -> 33 [ label = "" ]
    16 -> 34 [ label = "" ]
    16 -> 19 [ label = "" ]
    17 -> 32 [ label = "" ]
    17 -> 18 [ label = "" ]
    17 -> 35 [ label = "" ]
    18 -> 31 [ label = "" ]
    18 -> 32 [ label = "" ]
    18 -> 33 [ label = "" ]
    18 -> 36 [ label = "" ]
    18 -> 34 [ label = "" ]
    18 -> 16 [ label = "" ]
    18 -> 37 [ label = "" ]
    18 -> 19 [ label = "" ]
    20 -> 7 [ label = "" ]
    20 -> 25 [ label = "" ]
    29 -> 2 [ label = "" ]
    38 -> 3 [ label = "" ]
    38 -> 8 [ label = "" ]
    39 -> 3 [ label = "" ]
    39 -> 38 [ label = "" ]
    40 -> 3 [ label = "" ]
    21 -> 3 [ label = "" ]
    21 -> 38 [ label = "" ]
    22 -> 6 [ label = "" ]
    30 -> 10 [ label = "" ]
    30 -> 41 [ label = "" ]
    41 -> 10 [ label = "" ]
    41 -> 42 [ label = "" ]
    42 -> 10 [ label = "" ]
    42 -> 43 [ label = "" ]
    43 -> 10 [ label = "" ]
    26 -> 11 [ label = "" ]
    31 -> 11 [ label = "" ]
    31 -> 26 [ label = "" ]
    27 -> 12 [ label = "" ]
    27 -> 26 [ label = "" ]
    32 -> 12 [ label = "" ]
    32 -> 27 [ label = "" ]
    36 -> 13 [ label = "" ]
    34 -> 13 [ label = "" ]
    33 -> 13 [ label = "" ]
    33 -> 44 [ label = "" ]
    33 -> 45 [ label = "" ]
    33 -> 46 [ label = "" ]
    33 -> 47 [ label = "" ]
    33 -> 48 [ label = "" ]
    48 -> 13 [ label = "" ]
    48 -> 26 [ label = "" ]
    48 -> 27 [ label = "" ]
    47 -> 13 [ label = "" ]
    46 -> 13 [ label = "" ]
    46 -> 49 [ label = "" ]
    49 -> 13 [ label = "" ]
    49 -> 12 [ label = "" ]
    45 -> 13 [ label = "" ]
    44 -> 13 [ label = "" ]
    25 -> 15 [ label = "" ]
    25 -> 50 [ label = "" ]
    25 -> 51 [ label = "" ]
    51 -> 15 [ label = "" ]
    50 -> 15 [ label = "" ]
    50 -> 28 [ label = "" ]
    37 -> 16 [ label = "" ]
    28 -> 17 [ label = "" ]
    28 -> 35 [ label = "" ]
    35 -> 18 [ label = "" ]
    35 -> 37 [ label = "" ]
    24 -> 20 [ label = "" ]
    24 -> 52 [ label = "" ]
    52 -> 20 [ label = "" ]
    23 -> 20 [ label = "" ]
}

